tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,681 @@
+      // calls are connected to global op registry object so we should avoid
+      // going from one op registration call to another op registration call via
+      // the global registry object.
+      scanOpSchemaStrAndFunction(
+          I, instructions, &visitedOps, &visitedFunctions);
+      if (visitedOps.size() != 1) {
+        std::cerr << "[WARNING] found " << visitedOps.size() << " ops ( ";
+        for (auto& op : visitedOps) {
+          std::cerr << op << " ";
+        }
+        std::cerr << ") in a registration call in function: "
+                  << demangle(I->getFunction()->getName()) << std::endl;
+      }
+      for (const auto& op : visitedOps) {
+        opSchemaStrs->insert(op);
+        if (visitedFunctions.empty()) {
+          std::cerr << "[WARNING] could not find registered function for op: "
+                    << op << std::endl;
+        }
+        for (const auto& func : visitedFunctions) {
+          (*schemaStrToFunctions)[op].insert(func);
+          if (Verbose) {
+            std::cerr << "[DEBUG][OP_REG] " << op << " => "
+                      << demangle(func) << std::endl;
+          }
+        }
+      }
+    }
+  }
+
+  // Similar as scanOpRegistration - it searches for op schema strings that flow
+  // into c10 op invocation API call and assume the parent function of the API
+  // call invokes the operator.
+  //
+  // For example, from op name "aten::empty" to invocation API call:
+  // [OP_SCHEMA] aten::empty
+  // [PATH][1][CONST] [12 x i8] c"aten::empty\00"
+  // [PATH][2][CONST] @.str.69.1990 = private unnamed_addr constant [12 x i8]
+  //                  c"aten::empty\00", align 1
+  // [PATH][3][CONST] i8* getelementptr inbounds ([12 x i8], [12 x i8]*
+  //                  @.str.69.1990, i64 0, i64 0)
+  // [PATH][4][INST]  invoke void @std::basic_string<...>::basic_string(...
+  //                  (%"class.std::basic_string"* nonnull %19,
+  //                  i8* getelementptr inbounds ([12 x i8], [12 x i8]*
+  //                  @.str.69.1990, i64 0, i64 0) ...
+  // [PATH][5][INST]  %19 = alloca %"class.std::basic_string", align 8
+  // [PATH][6][INST]  %53 = bitcast %"class.std::basic_string"* %19 to i64*
+  // [PATH][7][INST]  %54 = load i64, i64* %53, align 8, !tbaa !4
+  // [PATH][8][INST]  store i64 %54, i64* %55, align 8, !tbaa !4
+  // [PATH][9][INST]  %55 = bitcast %"struct.c10::OperatorName"* %18 to i64*
+  // [PATH][10][INST] %18 = alloca %"struct.c10::OperatorName", align 8
+  // [PATH][11][INST] invoke void @c10::Dispatcher::findSchema(c10::OperatorName
+  //                  const&)(%"class.c10::optional.105"* nonnull sret %17,
+  //                  %"class.c10::Dispatcher.6320"* nonnull %45,
+  //                  %"struct.c10::OperatorName"* nonnull dereferenceable(16)
+  //                  %18)
+  static void scanOpInvocation(
+      VALUE_SET& instructions, SET* opSchemaStrs, GRAPH* functionToSchemaStrs) {
```

Comment:
The last argument here is called `functionToSchemaStrs` while at the previous function it's called `schemaStrToFunctions`, but we're passing the same graph into both of these fucntions - this probably needs to be cleaned up, and we need to make sure we're not accidentally reverse the meaning somewhere in the process.

Reply:
These functions are indeed populating different mappings (function -> op schema v.s. op schema -> function) so the naming here is proper. It's the decision we made at calling site to merge all these mappings into one graph (together with function -> function mapping) so it can perform end-to-end (op -> function -> op) analysis.
I added some comments at calling side to make it clear. Also explained different types of mappings above `simplifyGraph`.